packages/roosterjs-content-model-core/lib/coreApi/addUndoSnapshot/addUndoSnapshot.ts

```diff
@@ -11,19 +19,52 @@ import type { AddUndoSnapshot, Snapshot } from 'roosterjs-content-model-types';
  * when undo/redo to this snapshot
  */
 export const addUndoSnapshot: AddUndoSnapshot = (core, canUndoByBackspace, entityStates) => {
-    const { lifecycle, physicalRoot, undo } = core;
+    const { lifecycle, physicalRoot, logicalRoot, undo } = core;
     let snapshot: Snapshot | null = null;
 
     if (!lifecycle.shadowEditFragment) {
         // Need to create snapshot selection before retrieve innerHTML since HTML can be changed during creating selection when normalize table
         const selection = createSnapshotSelection(core);
         const html = physicalRoot.innerHTML;
 
+        // give plugins the chance to share entity states to include in the snapshot
+        const logicalRootEntityWrapper = logicalRoot.closest<HTMLElement>(`.${ENTITY_INFO_NAME}`);
+        if (!entityStates && logicalRootEntityWrapper) {
+            const entityFormat = parseEntityFormat(logicalRootEntityWrapper);
+            if (entityFormat.entityType && entityFormat.id) {
+                const event = <EntityOperationEvent>{
+                    eventType: 'entityOperation',
+                    operation: 'snapshotEntityState',
+                    entity: {
+                        type: entityFormat.entityType,
+                        id: entityFormat.id,
+                        wrapper: logicalRootEntityWrapper,
+                        isReadonly: entityFormat.isReadonly,
+                    },
+                    state: undefined,
+                };
+
+                core.api.triggerEvent(core, event, false);
+
+                // copy out any entity states from the plugins
+                if (event.state) {
+                    entityStates = [
+                        {
+                            type: entityFormat.entityType,
+                            id: entityFormat.id,
+                            state: event.state,
+                        },
+                    ];
+                }
+            }
+        }
+
         snapshot = {
             html,
             entityStates,
             isDarkMode: !!lifecycle.isDarkMode,
             selection,
+            logicalRootPath: getPath(logicalRoot, 0, physicalRoot),
```

Comment:
What is the value of logicalRootPath when logicalRoot==physicalRoot? I think it will be an empty array? Please double check.

We can make a perf optimization here. If we know they are the same, just directly write "[]" here and no need to call getPath()